service/src/main/java/dk/dbc/opensearch/reponse/SearchProcessorBean.java

```diff
@@ -108,21 +126,62 @@ private void process(SearchResponse response) throws XMLStreamException, IOExcep
         RepositorySettings repoSettings = getRepoSettings();
-        int start = request.getStartOrDefault();
-        int step = request.getStepValueOrDefault();
+        RepositoryAbstraction abstraction = repoSettings.abstraction();
         ResultSetKey key = ResultSetKey.of(request);
-        resultSet = resultSetCache.get(key);
-        if (resultSet == null)
-            resultSet = makeResultSet(key, repoSettings, trackingId);
+        getResultSet(key, repoSettings, trackingId, abstraction);
 
-        // Make it visible to others while we're working on it
-        // So that others don't make the same object
-        resultSetCache.put(key, resultSet);
-
-        resultSet.fetchWorks(repoSettings.solrClient(),
+        int start = request.getStartOrDefault();
+        int step = request.getStepValueOrDefault();
+        resultSet.fetchWorks(abstraction.getSolrClient(),
                              timings, start, step,
                              trackingId);
+        log.trace("resultSet = {}", resultSet);
+
         resultSetCache.put(key, resultSet);
+
+        fetchRecordsAndFormatThem(step, start, repoSettings, abstraction, trackingId);
+    }
+
+    private void fetchRecordsAndFormatThem(int step, int start, RepositorySettings repoSettings, RepositoryAbstraction abstraction, String trackingId) {
+        for (int i = 0 ; i < step ; i++) {
+            int index = start + i;
+            if (index > resultSet.getHitCount())
+                break;
+            String work = resultSet.workAtIndex(index);
+            List<String> units = resultSet.unitsForWork(work);
+            boolean formatThisUnit = true;
+            boolean formatMoreThanOne = formatMoreThanOneUnit();
+            for (String unit : units) {
+                List<String> openFormatFormatsForThisUnit =
+                        formatThisUnit ? getOpenFormatFormats(repoSettings) : EMPTY_LIST;
+
+                Future recordContent = es.submit(() -> abstraction.recordContent(
+                        fetcher, timings, trackingId,
+                        resultSet, request.getShowAgencyOrDefault(), unit,
+                        openFormatFormatsForThisUnit));
+                recordFecthing.put(unit, recordContent);
+                formatThisUnit = formatMoreThanOne;
+            }
+        }
+    }
+
+    private void getResultSet(ResultSetKey key, RepositorySettings repoSettings, String trackingId, RepositoryAbstraction abstraction) {
+        if (resultSetCache.putIfAbsent(key, EMPTY_RESULT_SET)) {
+            Profiles profiles = oaProfiles.getProfileFor(
+                    key.getAgencyId(), repoSettings.getName(),
+                    timings, trackingId, repoSettings.getSolrRules());
+            Profile profile = profiles.getProfile(key.getProfiles());
+            resultSet = abstraction.resultSetFor(key, profile);
+        } else {
+            resultSet = resultSetCache.get(key);
+        }
+    }
```

Comment:
_I may be wrong but_
```    private void getResultSet(ResultSetKey key, RepositorySettings repoSettings, String trackingId, RepositoryAbstraction abstraction) {
        if (resultSetCache.putIfAbsent(key, EMPTY_RESULT_SET)) {
            Profiles profiles = oaProfiles.getProfileFor(
                    key.getAgencyId(), repoSettings.getName(),
                    timings, trackingId, repoSettings.getSolrRules());
            Profile profile = profiles.getProfile(key.getProfiles());
            resultSet = abstraction.resultSetFor(key, profile);
        } else {
            resultSet = resultSetCache.get(key);
        }
    }
```
Vil dette ikke placere EMPTY_RESULT_SET i cachen og fetche den igen anden gang samme key benyttes.
Desuden vil compute() rutinen efterfølgende forsøge at opdatere EMPTY_RESULT_SET med data, mens nameOf* rutinerne returnerer en tom streng.


Er din tanke ikke snarere, at EMPTY_RESULT_SET er en factory der skaber et nyt tomt resultset i stedet for en ny nedarvet klasse.

Reply:
Sidst i metode bliver resultSet sat igen. Det er for at låse slottet i cache. Det kræver nok en kommentar eller 2.